setup.py

```diff
@@ -124,29 +124,18 @@
     "networkx.utils": ["tests/*.py"],
 }
 
-install_requires = ["decorator>=4.3.0"]
+
+def parse_requirements_file(filename):
+    with open(filename, encoding="utf-8") as fid:
+        requires = [l.strip() for l in fid.readlines() if l]
+
+    return requires
+
+
+install_requires = parse_requirements_file("requirements/default.txt")
 extras_require = {
-    "all": [
-        "numpy",
-        "scipy",
-        "pandas",
-        "matplotlib",
-        "pygraphviz",
-        "pydot",
-        "pyyaml",
-        "lxml",
-        "pytest",
-    ],
-    "gdal": ["gdal"],
-    "lxml": ["lxml"],
-    "matplotlib": ["matplotlib"],
-    "pytest": ["pytest"],
-    "numpy": ["numpy"],
-    "pandas": ["pandas"],
-    "pydot": ["pydot"],
-    "pygraphviz": ["pygraphviz"],
-    "pyyaml": ["pyyaml"],
-    "scipy": ["scipy"],
```

Comment:
[official guidance](https://setuptools.readthedocs.io/en/latest/userguide/dependency_management.html?highlight=extras_require#optional-dependencies) states that deprecating `extras` keys should be done by using empty lists (i.e. `"gdal": []`, etc.)

Reply:
Are you referring to the Note that begins: `Best practice: if a project ends up not needing any other packages to support a feature,`?  I don't believe that applies.  We are no longer supporting that feature.  The example describes a situation where a package used to require a third party package to support a feature, but then implements that support themselves.

If you need geospatial support, please see the geospatial gallery examples for the recommended way to work with geospatial data:  https://networkx.org/documentation/latest/auto_examples/index.html#geospatial

Reply:
Also, if you know of any packages that install `networkx[gdal]`, please let us know.

Reply:
Yes that's the note. It definitely applies.

A package which depends on e.g. `networkx[all]` will fail to install if this PR is released.

you HAVE to do `"all": [], "gdal": [], "lxml": [], "matplotlib": [], ...` there's no room for debate.

Reply:
Can you give a short recipe that shows the failure?

Such a failure doesn't make sense the way I understand how this works. 
But maybe I don't understand something...  Would this work?

```
pip install decorator
pip install networkx --no-deps
```

Reply:
As far as I can tell, pip works as expected according to the documentation:
```
$ pip install networkx[gdal]==2.6rc1
Collecting networkx[gdal]==2.6rc1
  Using cached networkx-2.6rc1-py3-none-any.whl (1.9 MB)
WARNING: networkx 2.6rc1 does not provide the extra 'gdal'
Collecting pandas>=1.1
  Using cached pandas-1.2.4-cp39-cp39-manylinux1_x86_64.whl (9.7 MB)
Collecting numpy>=1.19
  Using cached numpy-1.20.3-cp39-cp39-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (15.4 MB)
Collecting scipy!=1.6.1,>=1.5
  Using cached scipy-1.6.3-cp39-cp39-manylinux1_x86_64.whl (27.3 MB)
Collecting decorator<6,>=5.0.7
  Using cached decorator-5.0.9-py3-none-any.whl (8.9 kB)
Collecting matplotlib>=3.3
  Using cached matplotlib-3.4.2-cp39-cp39-manylinux1_x86_64.whl (10.3 MB)
Collecting kiwisolver>=1.0.1
  Using cached kiwisolver-1.3.1-cp39-cp39-manylinux1_x86_64.whl (1.2 MB)
Collecting pyparsing>=2.2.1
  Using cached pyparsing-2.4.7-py2.py3-none-any.whl (67 kB)
Collecting cycler>=0.10
  Using cached cycler-0.10.0-py2.py3-none-any.whl (6.5 kB)
Collecting pillow>=6.2.0
  Using cached Pillow-8.2.0-cp39-cp39-manylinux1_x86_64.whl (3.0 MB)
Collecting python-dateutil>=2.7
  Using cached python_dateutil-2.8.1-py2.py3-none-any.whl (227 kB)
Collecting six
  Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
Collecting pytz>=2017.3
  Using cached pytz-2021.1-py2.py3-none-any.whl (510 kB)
Installing collected packages: six, pytz, python-dateutil, pyparsing, pillow, numpy, kiwisolver, cycler, scipy, pandas, matplotlib, decorator, networkx
Successfully installed cycler-0.10.0 decorator-5.0.9 kiwisolver-1.3.1 matplotlib-3.4.2 networkx-2.6rc1 numpy-1.20.3 pandas-1.2.4 pillow-8.2.0 pyparsing-2.4.7 python-dateutil-2.8.1 pytz-2021.1 scipy-1.6.3 six-1.16.0
```

It installs NetworkX without error, but does correctly warn that we no longer provide the extra `gdal`.

Reply:
I also created a Python package that depended on `networkx[gdal]==2.6rc1` and it also installed without error.  But did emit the desired warning.

Reply:
Interesting. AFAIK it should've been an `ERROR`, not a `WARNING`. Seems odd to merely warn rather than error on an explicitly requested extra; my guess is `pip` did this to "avoid more issues than it creates" though I wouldn't trust this behaviour to remain indefinitely.

Up to you to keep as-is if you wanna.